phase2/kubeadm/do

```diff
@@ -0,0 +1,87 @@
+#!/bin/bash
+
+set -o errexit
+set -o pipefail
+set -o nounset
+set -x
+
+cd "${BASH_SOURCE%/*}"
+
+CONFIG_FILE="${CONFIG_JSON_FILE:-../../.config.json}"
+KUBEADM_INIT_PARAM_FILE=/etc/kubeadm/kubeadm_init_params.txt
+
+upgrade_master() {
+  KUBERNETES_VERSION=$(jq -r '.phase2.kubernetes_version' ${CONFIG_FILE})
+  UPGRADE_METHOD=$(jq -r '.phase2.kubeadm.master_upgrade.method' ${CONFIG_FILE})
+  MASTER=$(do_phase1 master-host)
+
+  update_host_kubeadm ${MASTER}
+  wait_nodes_ready
+
+  INIT_PARAM=$(execute_host ${MASTER} "sudo cat ${KUBEADM_INIT_PARAM_FILE}")
+  case "${UPGRADE_METHOD}" in
+    "init")
+      execute_host ${MASTER} "sudo kubeadm init --skip-preflight-checks --kubernetes-version ${KUBERNETES_VERSION} ${INIT_PARAM}"
+      ;;
+    "upgrade")
+      execute_host ${MASTER} "sudo kubeadm config upload from-flags ${INIT_PARAM}"
+      execute_host ${MASTER} "sudo kubeadm upgrade apply ${KUBERNETES_VERSION} -y -f"
+      ;;
+  esac
+}
+
+wait_nodes_ready() {
+  for TRIES in {1..60}; do
+    NODE_COUNT=$(kubectl get node --no-headers | wc -l)
+    READY_NODE_COUNT=$(kubectl get node --no-headers | grep -c '\sReady\s')
+    if [[ $NODE_COUNT -eq $READY_NODE_COUNT ]]; then
+      echo "All ${NODE_COUNT} nodes ready."
+      return
+    else
+      echo "${READY_NODE_COUNT}/${NODE_COUNT} nodes ready. Waiting for all nodes to become ready... ${TRIES}/60"
+      sleep 5
+    fi
+  done
+  echo "Gave up waiting for nodes to become ready." >&2
+  exit 1
+}
+
+update_host_kubeadm() {
+  HOST=$1 
+  KUBEADM_VERSION=$(jq -r '.phase2.kubeadm.version' ${CONFIG_FILE})
+  KUBELET_VERSION=$(jq -r '.phase2.kubelet_version' ${CONFIG_FILE})
+  if [[ "${KUBEADM_VERSION}" == stable ]]; then
+      # TODO(jessicaochen): Figure out how to translate "stable" to the corresponding GS link so we can
+      # update kubadm without disturbing the kubelet. Installing kubeadm via package will modify the kubelet configuration.
+      echo "Kubeadm version of 'stable' is not yet supported for upgrades. Please specify a Google Cloud Storage Link."
+      exit 1
+  elif [[ "${KUBEADM_VERSION}" == "gs://"* ]]; then
+    TMPDIR=/tmp/k8s-debs
+    execute_host ${HOST} "sudo mkdir ${TMPDIR}"
```

Comment:
Of these four execute_host commands, only the third (copying into /usr/bin) actually looks like it needs to be run under sudo, yet all of them are.

Reply:
You are probably right. I will revisit later.